### PR TITLE
Cisco email security integration

### PR DESCRIPTION
**Pull request by ChanochShayner**
## XSOAR [Partner](https://xsoar.pan.dev/docs/partners/become-a-tech-partner)?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**.

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/26215
## Description
integration for Cisco Email Security.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+            url_params += f'&spamPositive={argToBoolean(value)}'
+        if key == 'quarantined_as_spam' and value == 'True':
+            url_params += f'&quarantinedAsSpam={argToBoolean(value)}'
+        if key == 'virus_positive' and value == 'True':
+            url_params += f'&virusPositive={argToBoolean(value)}'
+        if key == 'contained_malicious_urls' and value == 'True':
+            url_params += f'&containedMaliciousUrls={argToBoolean(value)}'
+        if key == 'contained_neutral_urls' and value == 'True':
+            url_params += f'&containedNeutralUrls={argToBoolean(value)}'
+
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:176]: let's discuss this structure

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        if key == 'macro_file_types_detected':
+            url_params += f'&macroFileTypesDetected={value}'
+
+    return url_params
+
+
+def messages_to_human_readable(messages):
+    messages_readable_outputs = []
+    for message in messages:
+        readable_output = assign_params(message_id=message.get('attributes').get('mid'),
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:200]: it's for readable output.

File: Packs/CiscoEmailSecurity/pack_metadata.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "CiscoEmailSecurity",
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/pack_metadata.json:2]:
```suggestion
    "name": "Cisco Email Security",
```

File: Packs/CiscoEmailSecurity/pack_metadata.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    "description": "integration for Cisco Email Security",
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/pack_metadata.json:3]: update

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+  required: true
+  type: 0
+- display: API Password
+  hidden: false
+  name: api_password
+  required: true
+  type: 4
+description: This is the Hello World integration for getting started.
+display: CiscoEmailSecurity
+name: CiscoEmailSecurity
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:70]:
```suggestion
name: Cisco Email Security
```
in the display and name as well

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        elif key == 'device_group':
+            url_params += f'&{key}={value}'
+        elif key == 'device_name':
+            url_params += f'&{key}={value}'
+
+    return url_params
+
+
+def list_report_command(client: Client, args: Dict[str, Any]):
+    url_suffix = '/api/v2.0/reporting'
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:121]: why isnt the path in the client class

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+
+
+def messages_to_human_readable(messages):
+    messages_readable_outputs = []
+    for message in messages:
+        readable_output = assign_params(message_id=message.get('attributes').get('mid'),
+                                        cisco_message_id=message.get('attributes').get('icid'),
+                                        sender=message.get('attributes').get('sender'),
+                                        sender_ip=message.get('attributes').get('senderIp'),
+                                        subject=message.get('attributes').get('subject'),
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:203]:
```suggestion
                                        subject=message.get('attributes', {}).get('subject'),
```

**Review by ronykoz — COMMENTED** [review]
update the image of the integration please

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+class Client(BaseClient):
+
+    def __init__(self, params):
+        self.username = params.get('api_username')
+        self.password = params.get('api_password')
+        super().__init__(base_url=params.get('base_url'), verify=not params.get('insecure', False),
+                         ok_codes=tuple(), proxy=params.get('proxy', False))
+
+        # self._token_base64 = self._generate_base64_token()
+        # self._headers = {'Authorization': 'Basic ' + self._token_base64}
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:27]: delete this

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+
+        self._jwt_token = self._generate_jwt_token()
+        self._headers = {
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+            'jwtToken': self._jwt_token
+        }
+
+    def http_request(self, method, url_suffix, full_url=None, headers=None, json_data=None, params=None, data=None,
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:36]: why do you need this if all you are doing is super

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+    def list_entries_delete(self, url_suffix, request_body):
+        return self.http_request(
+            method='DELETE',
+            url_suffix=url_suffix,
+            json_data=request_body
+        )
+
+
+def test_module(client: Client) -> str:
+    suffix_url = "/sma/api/v2.0/message-tracking/messages?startDate=2018-01-01T00:00:00.000Z&" \
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:148]: why are we mentioning dates? what happens if the date doesn't exist in the system?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+        jwt_token = output.get('data').get('jwtToken')
+        return jwt_token
+
+    def list_report(self, url_suffix) -> Dict[str, Any]:
+        return self.http_request(
+            method='GET',
+            url_suffix=url_suffix,
+        )
+
+    def list_messages(self, url_suffix) -> Dict[str, Any]:
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:70]: the url suffix shouldn't be as part of the request, you should have it built-in.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+    return url_params
+
+
+def list_report_command(client: Client, args: Dict[str, Any]):
+    url_suffix = '/sma/api/v2.0/reporting'
+    url_params = build_url_params_for_list_report(args)
+    url_suffix_to_filter_by = url_suffix + url_params
+    report_response_data = client.list_report(url_suffix_to_filter_by)
+    return CommandResults(
+        readable_output='human_readable',
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:200]: Human readable?
You should give a proper header or just not mention anything

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+        outputs_prefix='CiscoEmailSecurity.report',
+        outputs_key_field='',
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:202]: there is no key?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+        outputs_key_field='messages.mid',
+        outputs=message_data
+    )
+
+
+def list_get_url_details_command(client, args):
+    url_suffix_to_filter_by = build_url_params_for_get_details(args, '/sma/api/v2.0/message-tracking/url-details')
+    message_get_details_response_data = client.list_get_url_details(url_suffix_to_filter_by)
+    message_data, human_readable = response_data_to_context_and_human_readable(message_get_details_response_data)
+    return CommandResults(
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:373]: Shouldn't you create URL indicator context as well?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 15]
```diff
@@ -0,0 +1,1156 @@
...
+configuration:
+- display: Server URL (e.g. https://192.168.0.1)
+  name: base_url
+  required: true
+  type: 0
+- display: Fetch incidents
+  name: isFetch
+  required: false
+  type: 8
+- display: Incident type
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:14]: where is the fetch in the code?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 16]
```diff
@@ -0,0 +1,1156 @@
+category: Utilities
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:1]: Don't think it's the right category

**Review by ronykoz — CHANGES_REQUESTED** [review]
1. Update the image
2. Is the report comes out as a file? if so let's save it the as such
3. You are not mentioning that this is a beta, please fix that.
4. Fix the URL passing to the client functions

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 10 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:148]: The dates are a required parameter for all API queries

Comment (ronykoz) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:148]: hat happens if the date doesn't exist in the system?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 13 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:202]: This command - `cisco-email-security-report-get` doesn't exist in swagger, and in the API docs it appears without any example - so its implementation is not clear.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity_test.py [hunk 17]
```diff
@@ -171,13 +303,31 @@ def test_list_release_quarantine_messages_command(requests_mock):
 
 
 def test_build_url_filter_for_get_list_entries():
+    """
+    Given:
+        Arguments To flirt with.
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity_test.py:308]: filter

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 18]
```diff
@@ -57,96 +42,107 @@ def _generate_jwt_token(self) -> str:
...
-            url_suffix=url_suffix,
+            url_suffix=f"/sma/api/v2.0/quarantine/{list_type}",
             json_data=request_body
         )
 
 
+def get_dates_for_test_module():
+    now = datetime.now()
+    start = now + timedelta(days=1)
+    end = now + timedelta(days=2)
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:135]: should this be in the future?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 19]
```diff
@@ -192,12 +188,10 @@ def build_url_params_for_list_report(args):
...
 
 def list_report_command(client: Client, args: Dict[str, Any]):
-    url_suffix = '/sma/api/v2.0/reporting'
     url_params = build_url_params_for_list_report(args)
-    url_suffix_to_filter_by = url_suffix + url_params
-    report_response_data = client.list_report(url_suffix_to_filter_by)
+    report_response_data = client.list_report(url_params)
     return CommandResults(
-        readable_output='human_readable',
+        readable_output='',
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:194]: why don't we have readable output here?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 14 above)
Comment (ronykoz) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:363]: @ChanochShayner what about this one?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 20]
```diff
@@ -131,8 +131,8 @@ def list_entries_delete(self, list_type, request_body):
 
 def get_dates_for_test_module():
     now = datetime.now()
-    start = now + timedelta(days=1)
-    end = now + timedelta(days=2)
+    start = now - timedelta(days=1)
+    end = now - timedelta(days=2)
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:135]: start should be before the end

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 21]
```diff
@@ -0,0 +1,1220 @@
...
+  required: true
+  type: 0
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:17]: bring this two to the bottom of the params

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 22]
```diff
@@ -0,0 +1,639 @@
...
+
+def list_report_command(client: Client, args: Dict[str, Any]):
+    counter = args.get('report_counter')
+    url_params = build_url_params_for_list_report(args, counter)
+    report_response_data = client.list_report(url_params)
+    report_data = report_response_data.get('data', {}).get('resultSet', [None])[0]
+    counter_output_prefix = set_var_to_output_prefix(counter)
+    return CommandResults(
+        readable_output=f'{report_response_data}',
+        outputs_prefix=f'CiscoEmailSecurity.Report.{counter_output_prefix}',
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:216]: why is the prefix being randomized?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 14 above)
Comment (ronykoz) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:384]: @ChanochShayner any update regarding the URL indicator?

**Review by ronykoz — COMMENTED** [review]
Don't we need a description here?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 22 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:216]: In report command, there are five things to report and I thought everyone needed a different path in context, I'm not right?
Should everyone be under report?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 14 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:384]: The fields in response data that returns do not match the fields in the URL context standard,
I asked @Arsenikr to check this.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 23]
```diff
@@ -0,0 +1,1220 @@
...
+- display: API Username
+  hidden: false
+  name: api_username
+  required: true
+  type: 0
+- display: API Password
+  hidden: false
+  name: api_password
+  required: true
+  type: 4
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:27]: why not just use credentials?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 24]
```diff
@@ -0,0 +1,1220 @@
...
+  - arguments:
+    - default: false
+      description: Date from which the search begins. Template for date - YYYY-MM-DD
+        hh:mm:00
+      isArray: false
+      name: start_date
+      required: true
+      secret: false
+    - default: false
+      description: Date the search ends. Template for date - YYYY-MM-DD hh:mm:00
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:43]: Please add a note stating that the seconds are ignored, due to API limits, here and everywhere

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 25]
```diff
@@ -0,0 +1,1220 @@
...
+    - default: false
+      description: Address to filter with.
+      isArray: false
+      name: order_by_from_address
+      required: false
+      secret: false
+    - default: false
+      description: Address to filter with.
+      isArray: false
+      name: order_by_to_address
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:629]: The two descriptions are the same

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 26]
```diff
@@ -0,0 +1,1220 @@
...
+      description: Subject to filter with.
+      isArray: false
+      name: order_by_subject
+      required: false
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      description: Address order filtering.
+      isArray: false
+      name: order_dir_from_address
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:642]: unclear name

**Review by guykeller — CHANGES_REQUESTED** [review]

**Comment by guykeller** [comment]
@ChanochShayner 
Looks good, some general notes:
1. Let's make sure the names are clear
2. Make sure that context paths are always in singular form, unless there's a reason not to  make them so
3. If any of the arguments for commands is an array, please verify it has `isArray: true`

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 27]
```diff
@@ -0,0 +1,639 @@
...
+    human_readable = tableToMarkdown('CiscoEmailSecurity Messages', messages_readable_outputs, headers, removeNull=True)
+    return human_readable
+
+
+def list_search_messages_command(client, args):
+    url_params = build_url_params_for_list_messages(args)
+    messages_response_data = client.list_messages(url_params)
+    messages_data = messages_response_data.get('data')
+    for message in messages_data:
+        message['attributes']['mid'] = message.get('attributes', {}).get('mid', [None])[0]
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:306]: The message ID is in a list, and in order to arrange the context and the readable output, I remove it from the list -
image

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 28]
```diff
@@ -0,0 +1,639 @@
...
+    }
+
+    if args.get('recipient_addresses'):
+        request_body["recipientAddresses"] = args.get('recipient_addresses').split(',')
+    if args.get('recipient_list'):
+        request_body["recipientList"] = args.get('recipient_list').split(',')
+    if args.get('sender_addresses'):
+        request_body["senderAddresses"] = args.get('sender_addresses').split(',')
+    if args.get('sender_list'):
+        request_body["senderList"] = args.get('sender_list').split(',')
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:541]: There is a choice between the lists and the addresses -
You can select one list and one addresses
So it seems impossible to use square brackets.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 29]
```diff
@@ -0,0 +1,639 @@
...
+    return request_body
+
+
+def set_outputs_key_field_for_list_entries(args):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:545]: The key field for the context depends on the customer choice  - 
Recipient list or Sender list

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 30]
```diff
@@ -0,0 +1,639 @@
...
+def list_entries_delete_command(client, args):
+    list_type = args.get('list_type')
+    request_body = build_request_body_for_delete_list_entries(args)
+    list_entries_response = client.list_entries_delete(list_type, request_body)
+    list_entries = list_entries_response.get('data')
+    output_prefix = list_type.title()
+    outputs_key_field = set_outputs_key_field_for_list_entries(args)
+    return CommandResults(
+        readable_output=list_entries,
+        outputs_prefix=f'CiscoEmailSecurity.listEntries.{output_prefix}',
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:584]: A different path for a safe list and block list

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 31]
```diff
@@ -0,0 +1,639 @@
...
+
+
+def build_url_params_for_list_report(args, report_counter):
+    start_date = date_to_cisco_date(args.get('start_date'))
+    end_date = date_to_cisco_date(args.get('end_date'))
+    url_params = f'/{report_counter}?startDate={start_date}&endDate={end_date}&device_type=esa'
+    return url_params
+
+
+def set_var_to_output_prefix(report_counter):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:198]: This method gets a variable and returns it according to the standard of outputs prefix - 
mail_incoming_traffic_summary -> MailIncomingTrafficSummary

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 14 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:384]: According to @Arsenikr - There is no need to do this.
We align according to this standard only if we make a command !Url and treat it as an indicator.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (guykeller) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:323]: So let's split it into several lines, for example:
message_details = message.get('attributes', {})
message_id = message_details.get('mid', [None])[0]
message['attributes']['mid'] = message_id

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py [hunk 32]
```diff
@@ -0,0 +1,654 @@
...
+        "viewBy": args.get('view_by')
+    }
+
+    if args.get('recipient_addresses'):
+        request_body["recipientAddresses"] = args.get('recipient_addresses').split(',')
+    if args.get('recipient_list'):
+        request_body["recipientList"] = args.get('recipient_list').split(',')
+    if args.get('sender_addresses'):
+        request_body["senderAddresses"] = args.get('sender_addresses').split(',')
+    if args.get('sender_list'):
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:555]: those checks (if statement conditions) should be:
if 'recipient_list' in args
Shouldn't it?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (guykeller) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:560]: so let's give it a better name + a docstring if necessary

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml (hunk 26 above)
Comment (guykeller) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:644]: @ChanochShayner Did you see this?

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml [hunk 33]
```diff
@@ -0,0 +1,1224 @@
...
+      isArray: false
+      name: macro_file_types_detected
+      required: false
+      secret: false
+    deprecated: false
+    description: Cisco Email Security search messages.
+    execution: false
+    name: cisco-email-security-messages-search
+    outputs:
+    - contextPath: CiscoEmailSecurity.Messages.attributes.direction
```
Comment (guykeller) [inline Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:426]: 'Messages' is still in plural, and also some other context outputs have the same problem.
Let's fix this please

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:323]: Done

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 32 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:555]: You right, done.

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.py:560]: done

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml (hunk 33 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:426]: done

File: Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml (hunk 26 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/Integrations/CiscoEmailSecurity/CiscoEmailSecurity.yml:644]: Done, From another test I did, according to the swagger, I changed the above arguments.

**Comment by kirbles19** [comment]
@ChanochShayner Done.

**Review by guykeller — APPROVED** [review]

File: Packs/CiscoEmailSecurity/pack_metadata.json [hunk 34]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Cisco Email Security (Beta)",
```
Comment (ronykoz) [inline Packs/CiscoEmailSecurity/pack_metadata.json:2]: not sure if the name of the pack should contain beta, please double check

**Review by ronykoz — APPROVED** [review]

File: Packs/CiscoEmailSecurity/pack_metadata.json (hunk 34 above)
Comment (ChanochShayner) [reply Packs/CiscoEmailSecurity/pack_metadata.json:2]: I checked, and it seems it should contain beta

**Pull request merged**